src/jit/optcse.cpp

```diff
@@ -920,79 +968,251 @@ void Compiler::optValnumCSE_Availablity()
+#endif
+                    // Have we decided to abandon work on this CSE?
+                    if (desc->defExcSetPromise == ValueNumStore::NoVN)
                     {
-                        /* This is a CSE use */
+                        // This candidate had defs with differing liberal exc set VNs
+                        // We have abandoned CSE promotion for this candidate
 
-                        desc->csdUseCount += 1;
-                        desc->csdUseWtCnt += stmw;
+                        // Clear the CSE flag
+                        tree->gtCSEnum = NO_CSE;
+
+                        JITDUMP(" Abandoned - CSE candidate has defs with different exception sets!\n");
+                        continue;
                     }
-                    else
+
+                    // Record the exception set for tree's liberal value number
+                    //
+                    ValueNum theLiberalExcSet = vnStore->VNExceptionSet(tree->gtVNPair.GetLiberal());
+
+                    // Is this a CSE use or a def?
+
+                    if (isDef)
                     {
+                        // @ToDo - Remove this block as it no longer applies
                         if (tree->gtFlags & GTF_COLON_COND)
                         {
                             // We can't create CSE definitions inside QMARK-COLON trees
                             tree->gtCSEnum = NO_CSE;
+
+                            JITDUMP(" NO_CSE - This CSE def occurs in a GTF_COLON_COND!\n");
                             continue;
                         }
 
-                        /* This is a CSE def */
+                        // This is a CSE def
+
+                        // Is defExcSetCurrent still set to the uninit marker value of VNForNull() ?
+                        if (desc->defExcSetCurrent == vnStore->VNForNull())
+                        {
+                            // This is the first time visited, so record this defs exeception set
+                            desc->defExcSetCurrent = theLiberalExcSet;
+                        }
+
+                        // Have we seen a CSE use and made a promise of an exception set?
+                        //
+                        if (desc->defExcSetPromise != vnStore->VNForEmptyExcSet())
+                        {
+                            // The exeception set held in desc->defExcSetPromise must be a subset of theLiberalExcSet
+                            //
+                            if (vnStore->VNExcIsSubset(theLiberalExcSet, desc->defExcSetPromise))
+                            {
+                                // This new def still satisfies any promise made to all the CSE uses that we have
+                                // encountered
+                                //
+                                if (desc->defExcSetCurrent !=
+                                    theLiberalExcSet) // no update is needed when these are the same VN
+                                {
```

Comment:
I think the code reads better if you don't use these end of line comments.

Also it would be nice to try and bubble all the fail/continue cases upwards so that we don't get so deeply nested, eg structure the code as the loop-equivalent of early return style, eg:
```C++
  if (something disqualifying 1)
  { 
         continue;
  }

  if (something disqualifying 2)
  { 
         continue;
  }

  // success !
```

Reply:
Not sure exactly you want here. As far as I can tell this isn't a straight forward change.

Reply:
For now maybe just move the comments.

I think there may be simpler/cleaner ways to express the code too, but I'm ok leaving it as is for now.